lib/services/notification-api-service.js

```diff
@@ -0,0 +1,95 @@
+// Copyright 2016, EMC, Inc.
+
+'use strict';
+
+var di = require('di');
+
+module.exports = notificationApiServiceFactory;
+di.annotate(notificationApiServiceFactory, new di.Provide('Http.Services.Api.Notification'));
+di.annotate(notificationApiServiceFactory,
+    new di.Inject(
+        'Protocol.Events',
+        'Protocol.Task',
+        'TaskGraph.Store',
+        'Logger',
+        'Services.Waterline',
+        'Errors',
+        'Promise',
+        '_'
+    )
+);
+
+function notificationApiServiceFactory(
+    eventsProtocol,
+    taskProtocol,
+    taskGraphStore,
+    Logger,
+    waterline,
+    Errors,
+    Promise,
+    _
+) {
+    var logger = Logger.initialize(notificationApiServiceFactory);
+
+    function notificationApiService() {
+    }
+
+    notificationApiService.prototype.postNotification = function(message) {
+        var self = this;
+        var activeTaskId, activeGraphId;
+        var nodeId;
+
+        return Promise.try(function() {
+            if (!message.nodeId || !_.isString(message.nodeId)) {
+                throw new Errors.BadRequestError('Missing node ID in query or body');
+            };
+            if (!message.type || !_.isString(message.type)) {
+                throw new Errors.BadRequestError('Missing notification type in query or body');
+            };
+            if (!message.data || !_.isString(message.data)) {
+                throw new Errors.BadRequestError('Missing notification data in query or body');
+            }            
+        })
+        .then(function() {
+            nodeId = message.nodeId;
+            // Check if the node exists
+            return waterline.nodes.needByIdentifier(nodeId)
+        })
+        .then(function () {
+            return [
+                taskGraphStore.findActiveGraphForTarget(nodeId),
+                taskProtocol.activeTaskExists(nodeId)
+            ]
+        })
+        .spread(function (activeGraph, activeTask) {
```

Comment:
Since the activeGraph and activeTask is got by looking up the active task not passed from notification request.  There is a chance that the notification isn't the task wants.

Assume there is a Task A, it will create a thread in Microkernel, then it will post notification to tell RackHD it has finished.  In a case, the thread is hang for a long time (but actually it is still alive in microkernel), then RackHD forces to kill Task A, and retry. Another thread will be created in Microkernel, since previous thread is still alive, it finally success and sends the notification.  The new Task will receive the notification but actually this is not the notification it wants.

So I think it is still required to pass the taskId in notification request. If the taskId from notification doesn't match with the active task, then the notification should be ignored (we can log some warning here)